### PR TITLE
[GRPC] Add structured references for filters + GRPC batch delete endpoint

### PR DESCRIPTION
**Pull request by dirkkul**
### What's being changed:

- Adds structured references instead of paths (`["reference_property, "target_class", "property_in_target_class"]`) to filters in GRPC. 
- As batch_delete also uses filters, this PR also adds a GRPC batch_delete endpoint which reuses the existing filter extraction code.

This allows for a simpler UX on the client side as the target_class can be looked up in the schema (unless it is a multi-target reference) and does not need to be supplied by users.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/handlers/grpc/v1/service.go [hunk 1]
```diff
@@ -49,6 +49,39 @@ func NewService(traverser *traverser.Traverser, authComposer composer.TokenFunc,
...
+	principal, err := s.principalFromContext(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("extract auth: %w", err)
+	}
+	replicationProperties := extractReplicationProperties(req.ConsistencyLevel)
+	scheme := s.schemaManager.GetSchemaSkipAuth()
+
+	params, err := batchDeleteParamsFromProto(req, scheme)
+	if err != nil {
+		return nil, err
```
Comment (tsmith023) [inline adapters/handlers/grpc/v1/service.go:63]: Can we wrap this error (and other errors in this function) to show the location in which they were thrown? E.g. `fmt.Errorf("batch delete: %w", err)` and likewise for other errors in this PR

Actually, I feel this should be the same for all the methods in the gRPC API.
Perhaps we can do them all in another PR?

File: grpc/proto/v1/batch_delete.proto [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  int64 matches = 3;
+  int64 successful = 4;
+  repeated Object objects = 5;
+}
+
+message Object {
+  bytes uuid = 1;
+  bool successful = 2;
+  optional string error = 3;  // empty string means no error
+}
```
Comment (tsmith023) [inline grpc/proto/v1/batch_delete.proto:32]: I think the name `Object` here is too general for its application in the specific use as the `objects` of `BatchDeleteReply`. Can we name it equivalently to the core convention of `BatchSimpleObject` instead?

File: usecases/objects/batch_delete.go [hunk 3]
```diff
@@ -46,6 +46,28 @@ func (b *BatchManager) DeleteObjects(ctx context.Context, principal *models.Prin
...
 }
 
+// DeleteObjectsFromGRPC deletes objects in batch based on the match filter
+func (b *BatchManager) DeleteObjectsFromGRPC(ctx context.Context, principal *models.Principal,
+	params BatchDeleteParams,
+	repl *additional.ReplicationProperties, tenant string,
+) (BatchDeleteResult, error) {
+	err := b.authorizer.Authorize(principal, "delete", "batch/objects")
+	if err != nil {
+		return BatchDeleteResult{}, err
```
Comment (tsmith023) [inline usecases/objects/batch_delete.go:56]: Can we wrap the error here to log the function it was called in?

File: grpc/proto/v1/base.proto [hunk 4]
```diff
@@ -54,3 +54,81 @@ message ObjectProperties {
...
+    string value_text = 4;
+    int64 value_int = 5;
+    bool value_boolean = 6;
+    double value_number = 7;
+    TextArray value_text_array = 9;
+    IntArray value_int_array = 10;
+    BooleanArray value_boolean_array = 11;
+    NumberArray value_number_array = 12;
+    GeoCoordinatesFilter value_geo = 13;
+  };
```
Comment (tsmith023) [inline grpc/proto/v1/base.proto:107]: Just a thought, is filtering on phone numbers a valid use-case?

File: grpc/proto/v1/base.proto [hunk 5]
```diff
@@ -54,3 +54,81 @@ message ObjectProperties {
...
+
+message FilterReferenceSingleTarget {
+  string on = 1;
+  FilterTarget target = 2;
+}
+
+message FilterReferenceMultiTarget {
+  string on = 1;
+  FilterTarget target = 2;
+  string string_target_collection = 3;
```
Comment (tsmith023) [inline grpc/proto/v1/base.proto:119]: Is the `string_` prefix required here since the type is defined as a string already?

**Review by tsmith023 — CHANGES_REQUESTED** [review]
I think renaming `Object` is the only thing that I would like changed, if you agree
The `err` wrapping can go in another PR so as to keep this one small(ish)

File: grpc/proto/v1/base.proto (hunk 4 above)
Comment (dirkkul) [reply grpc/proto/v1/base.proto:107]: pretty sure that this is not supported by weaviate itself. At least the docs do not mention it at all

File: adapters/handlers/grpc/v1/service.go (hunk 1 above)
Comment (dirkkul) [reply adapters/handlers/grpc/v1/service.go:63]: I added it in the service file, but don't want to add it in all subfunctions - that is just too much copy&paste and often not really needed

File: grpc/proto/v1/batch_delete.proto (hunk 2 above)
Comment (dirkkul) [reply grpc/proto/v1/batch_delete.proto]: used BatchDeleteObject

File: usecases/objects/batch_delete.go (hunk 3 above)
Comment (dirkkul) [reply usecases/objects/batch_delete.go:56]: I don't want to add it to every call :/

the call stacks are sometimes many many levels deep and adding too much redundant info drowns out the relevant parts.

File: grpc/proto/v1/base.proto (hunk 5 above)
Comment (dirkkul) [reply grpc/proto/v1/base.proto]: yeah, not needed. thx :)

File: usecases/objects/batch_delete.go (hunk 3 above)
Comment (tsmith023) [reply usecases/objects/batch_delete.go:56]: I've found that having the errors wrapped greatly aids when debugging as you can string search for where the error was thrown in the message. But you're right that the messages can get too long for many calls

**Review by tsmith023 — APPROVED** [review]

File: grpc/proto/v1/base.proto [hunk 6]
```diff
@@ -54,3 +54,81 @@ message ObjectProperties {
...
+  Operator operator = 1;
+  // protolint:disable:next REPEATED_FIELD_NAMES_PLURALIZED
+  repeated string on = 2  [deprecated = true];  // will be removed in the future, use path
+  repeated Filters filters = 3;
+  oneof test_value{
+    string value_text = 4;
+    int64 value_int = 5;
+    bool value_boolean = 6;
+    double value_number = 7;
+    TextArray value_text_array = 9;
```
Comment (antas-marcin) [inline grpc/proto/v1/base.proto:102]: should be 8 as a next number not 9

File: grpc/proto/v1/base.proto [hunk 7]
```diff
@@ -54,3 +54,81 @@ message ObjectProperties {
...
+    FilterReferenceSingleTarget single_target = 2;
+    FilterReferenceMultiTarget multi_target = 3;
+  };
+}
+
+message GeoCoordinatesFilter {
+  float latitude = 1;
+  float longitude = 2;
+  float distance = 3;
+}
```
Comment (antas-marcin) [inline grpc/proto/v1/base.proto:134]: let's a newline at the end

File: grpc/proto/v1/batch_delete.proto [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  int64 matches = 3;
+  int64 successful = 4;
+  repeated BatchDeleteObject objects = 5;
+}
+
+message BatchDeleteObject {
+  bytes uuid = 1;
+  bool successful = 2;
+  optional string error = 3;  // empty string means no error
+}
```
Comment (antas-marcin) [inline grpc/proto/v1/batch_delete.proto:32]: maybe let's add a newline at the end

File: grpc/proto/v1/base.proto (hunk 6 above)
Comment (dirkkul) [reply grpc/proto/v1/base.proto:102]: I can't change the numbers now, as that would break older clients - IIRC 8 was something that we didn't need in the end, so we removed it.

File: usecases/objects/batch_delete.go [hunk 9]
```diff
@@ -46,6 +46,28 @@ func (b *BatchManager) DeleteObjects(ctx context.Context, principal *models.Prin
...
+	}
+
+	unlock, err := b.locks.LockConnector()
+	if err != nil {
+		return BatchDeleteResult{}, NewErrInternal("could not acquire lock: %v", err)
+	}
+	defer unlock()
```
Comment (antas-marcin) [inline usecases/objects/batch_delete.go:63]: shouldn't we run `unlock()` method also in case of error?

Comment (dirkkul) [reply usecases/objects/batch_delete.go:63]: I copied that from the function above `DeleteObjects` - I assume we didn't get the lock if there was an error.

I quickly checked other usage of `m.locks.LockConnector()`and they all do the same

**Review by antas-marcin — APPROVED** [review]
